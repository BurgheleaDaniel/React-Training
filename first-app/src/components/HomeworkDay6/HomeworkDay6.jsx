import React, { useState, useReducer, useEffect } from "react"
import TextField from "@material-ui/core/TextField"
import styles from "./HomeworkDay6.module.css"
import {
  Button,
  Checkbox,
  FormControlLabel,
  MenuItem,
  Select,
} from "@material-ui/core"

const initialState = {
  firstName: "",
  lastName: "",
  email: "",
  password: "",
  country: "",
  terms: false,
}

const reducer = (state, action) => {
  if (action.name === "reset") return initialState
  return {
    ...state,
    [action.name]: action.value,
  }
}

const HomeworkDay6 = () => {
  const [formState, setFormState] = useReducer(reducer, initialState)
  const [isFormSubmiting, setIsFormSubmiting] = useState(false)
  const [errors, setErrors] = useState([])
  const resetFormState = () => {
    setFormState({
      name: "reset",
    })
  }

  useEffect(() => {
    if (errors.length > 0) {
      isFormValid()
    }
  }, [formState])

  const isFormValid = () => {
    const err = []

    for (const [name, value] of Object.entries(formState)) {
      if (value === initialState[name]) {
        err.push(name)
      }
      setErrors(err)
    }

    return errors.length === 0
  }

  const handleSubmit = e => {
    e.preventDefault()
    if (isFormValid()) {
      setIsFormSubmiting(true)

      setTimeout(() => {
        setIsFormSubmiting(false)
        resetFormState()
      }, 2000)
    }
  }

  const handleInputChange = e => {
    const value =
      e.target.type === "checkbox" ? e.target.checked : e.target.value

    setFormState({
      name: e.target.name,
      value: value,
    })
  }

  const { firstName, lastName, email, password, country, terms } = formState

  return (
    <div className={styles.wrapper}>
      <form onSubmit={handleSubmit} className={styles.form}>
        <h1>Create Account</h1>

        <fieldset className={styles.fieldset}>
          <label htmlFor="firstName">
            <p>First Name</p>
            <TextField
              id="firstName"
              name="firstName"
              variant="outlined"
              value={firstName}
              onChange={handleInputChange}
              error={errors.includes("firstName") ? true : false}
              className={styles.input}
              inputProps={{
                "data-testid": "firstName",
              }}
            />
            {errors.includes("firstName") && (
              <p className={`${styles.errorMessage}`}>This Field is required</p>
            )}
          </label>
        </fieldset>

        <fieldset className={styles.fieldset}>
          <label htmlFor="lastName">
            <p>Last Name</p>
            <TextField
              id="lastName"
              name="lastName"
              variant="outlined"
              value={lastName}
              onChange={handleInputChange}
              error={errors.includes("lastName") ? true : false}
              className={styles.input}
              inputProps={{
                "data-testid": "lastName",
              }}
            />
            {errors.includes("lastName") && (
              <p className={`${styles.errorMessage}`}>This Field is required</p>
            )}
          </label>
        </fieldset>

        <fieldset className={styles.fieldset}>
          <label htmlFor="email">
            <p>Email</p>
            <TextField
              id="email"
              name="email"
              type="email"
              variant="outlined"
              value={email}
              onChange={handleInputChange}
              error={errors.includes("email") ? true : false}
              className={styles.input}
              inputProps={{
                "data-testid": "email",
              }}
            />
            {errors.includes("email") && (
              <p className={`${styles.errorMessage}`}>This Field is required</p>
            )}
          </label>
        </fieldset>

        <fieldset className={styles.fieldset}>
          <label htmlFor="password">
            <p>Password</p>
            <TextField
              id="password"
              name="password"
              type="password"
              variant="outlined"
              value={password}
              onChange={handleInputChange}
              error={errors.includes("password") ? true : false}
              className={styles.input}
              inputProps={{
                "data-testid": "password",
              }}
            />
            {errors.includes("password") && (
              <p className={`${styles.errorMessage}`}>This Field is required</p>
            )}
          </label>
        </fieldset>

        <fieldset className={styles.fieldset}>
          <label htmlFor="country">
            <p>Country</p>
            <Select
              id="country"
              name="country"
              variant="outlined"
              value={country}
              onChange={handleInputChange}
              label="Country"
              className={styles.input}
              inputProps={{
                "data-testid": "country",
              }}
            >
              <MenuItem value="">
                <em>Select a Country</em>
              </MenuItem>
              <MenuItem value="RO">RO</MenuItem>
              <MenuItem value="ES">ES</MenuItem>
              <MenuItem value="AU">AU</MenuItem>
              <MenuItem value="UK">UK</MenuItem>
            </Select>
            {errors.includes("country") && (
              <p className={`${styles.errorMessage}`}>This Field is required</p>
            )}
          </label>
        </fieldset>

        <fieldset className={styles.fieldset}>
          <label htmlFor="terms">
            <FormControlLabel
              control={
                <Checkbox
                  checked={terms}
                  onChange={handleInputChange}
                  id="terms"
                  name="terms"
                  data-testid="terms"
                  color="primary"
                />
              }
              label="I accept the terms of the service Wrap"
            />
            {errors.includes("terms") && (
              <p className={`${styles.errorMessage}`}>This Field is required</p>
            )}
          </label>
        </fieldset>

        <Button
          type="submit"
          data-testid="submitButton"
          variant="contained"
          color="primary"
          disabled={isFormSubmiting || errors.length > 0}
          className={styles.input}
        >
          Submit
        </Button>
      </form>
    </div>
  )
}

export default HomeworkDay6
