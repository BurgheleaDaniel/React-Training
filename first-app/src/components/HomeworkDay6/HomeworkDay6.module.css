.wrapper {
  background-color: #418ef1;
}

.form {
  background-color: #fff;
  width: 500px;
  margin: 0 auto;
  padding: 5px 20px;
}

.fieldset {
  /* margin: 20px 0; */
  border-style: unset;
}

.input {
  /* margin: 20px 0; */
  width: 90%;
}

.error {
  border: solid 1px red;
}

.errorMessage {
  color: red;
}
