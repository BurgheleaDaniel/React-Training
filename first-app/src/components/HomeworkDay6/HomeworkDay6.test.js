import React from "react"
import { render } from "@testing-library/react"
import renderer from "react-test-renderer"
import userEvent from "@testing-library/user-event"
import HomeworkDay6 from "./HomeworkDay6"

describe("Create Account Form", () => {
  it("should match snapshot", () => {
    const tree = renderer.create(<HomeworkDay6 />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("can type inyto input fields", () => {
    const { getByTestId, getAllByRole, getByText, debug } = render(
      <HomeworkDay6 />
    )

    userEvent.type(getByTestId("firstName"), "Daniel")
    userEvent.type(getByTestId("lastName"), "Burghelea")
    userEvent.type(getByTestId("email"), "daniel.burghelea@webbeds.com")
    userEvent.type(getByTestId("password"), "111111")

    expect(getByTestId("firstName").value).toBe("Daniel")
    expect(getByTestId("lastName").value).toBe("Burghelea")
    expect(getByTestId("email").value).toBe("daniel.burghelea@webbeds.com")
    expect(getByTestId("password").value).toBe("111111")
  })
})
