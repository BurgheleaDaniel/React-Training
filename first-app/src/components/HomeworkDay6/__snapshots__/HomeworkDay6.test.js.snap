// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Account Form should match snapshot 1`] = `
<div
  className="wrapper"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <h1>
      Create Account
    </h1>
    <fieldset
      className="fieldset"
    >
      <label
        htmlFor="firstName"
      >
        <p>
          First Name
        </p>
        <div
          className="MuiFormControl-root MuiTextField-root input"
        >
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input"
              data-testid="firstName"
              disabled={false}
              id="firstName"
              name="firstName"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
              value=""
            />
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
              style={
                Object {
                  "paddingLeft": 8,
                }
              }
            >
              <legend
                className="PrivateNotchedOutline-legend-2"
                style={
                  Object {
                    "width": 0.01,
                  }
                }
              >
                <span
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "&#8203;",
                    }
                  }
                />
              </legend>
            </fieldset>
          </div>
        </div>
      </label>
    </fieldset>
    <fieldset
      className="fieldset"
    >
      <label
        htmlFor="lastName"
      >
        <p>
          Last Name
        </p>
        <div
          className="MuiFormControl-root MuiTextField-root input"
        >
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input"
              data-testid="lastName"
              disabled={false}
              id="lastName"
              name="lastName"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
              value=""
            />
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
              style={
                Object {
                  "paddingLeft": 8,
                }
              }
            >
              <legend
                className="PrivateNotchedOutline-legend-2"
                style={
                  Object {
                    "width": 0.01,
                  }
                }
              >
                <span
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "&#8203;",
                    }
                  }
                />
              </legend>
            </fieldset>
          </div>
        </div>
      </label>
    </fieldset>
    <fieldset
      className="fieldset"
    >
      <label
        htmlFor="email"
      >
        <p>
          Email
        </p>
        <div
          className="MuiFormControl-root MuiTextField-root input"
        >
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input"
              data-testid="email"
              disabled={false}
              id="email"
              name="email"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="email"
              value=""
            />
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
              style={
                Object {
                  "paddingLeft": 8,
                }
              }
            >
              <legend
                className="PrivateNotchedOutline-legend-2"
                style={
                  Object {
                    "width": 0.01,
                  }
                }
              >
                <span
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "&#8203;",
                    }
                  }
                />
              </legend>
            </fieldset>
          </div>
        </div>
      </label>
    </fieldset>
    <fieldset
      className="fieldset"
    >
      <label
        htmlFor="password"
      >
        <p>
          Password
        </p>
        <div
          className="MuiFormControl-root MuiTextField-root input"
        >
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input"
              data-testid="password"
              disabled={false}
              id="password"
              name="password"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="password"
              value=""
            />
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
              style={
                Object {
                  "paddingLeft": 8,
                }
              }
            >
              <legend
                className="PrivateNotchedOutline-legend-2"
                style={
                  Object {
                    "width": 0.01,
                  }
                }
              >
                <span
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "&#8203;",
                    }
                  }
                />
              </legend>
            </fieldset>
          </div>
        </div>
      </label>
    </fieldset>
    <fieldset
      className="fieldset"
    >
      <label
        htmlFor="country"
      >
        <p>
          Country
        </p>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root input"
          onClick={[Function]}
        >
          <div
            aria-haspopup="listbox"
            aria-labelledby="country"
            className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
            id="country"
            onBlur={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            role="button"
            tabIndex={0}
          >
            <span
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&#8203;",
                }
              }
            />
          </div>
          <input
            aria-hidden={true}
            className="MuiSelect-nativeInput"
            data-testid="country"
            name="country"
            onAnimationStart={[Function]}
            onChange={[Function]}
            tabIndex={-1}
            value=""
          />
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="PrivateNotchedOutline-legendLabelled-3"
            >
              <span>
                Country
              </span>
            </legend>
          </fieldset>
        </div>
      </label>
    </fieldset>
    <fieldset
      className="fieldset"
    >
      <label
        htmlFor="terms"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-5 MuiCheckbox-root MuiCheckbox-colorPrimary MuiIconButton-colorPrimary"
            data-testid="terms"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={null}
          >
            <span
              className="MuiIconButton-label"
            >
              <input
                checked={false}
                className="PrivateSwitchBase-input-8"
                data-indeterminate={false}
                id="terms"
                name="terms"
                onChange={[Function]}
                type="checkbox"
              />
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
          </span>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            I accept the terms of the service Wrap
          </span>
        </label>
      </label>
    </fieldset>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained input MuiButton-containedPrimary"
      data-testid="submitButton"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="submit"
    >
      <span
        className="MuiButton-label"
      >
        Submit
      </span>
    </button>
  </form>
</div>
`;
